File: src/Language/OSA/AppleEventDescriptor.m

#import "Foundation/Foundation.h"

BOOL inline_c_Language_OSA_AppleEventDescriptor_0_e20f668435c58697c00c42f4b07f11aec03f001c(NSAppleEventDescriptor * evt_inline_c_0) {
return ( (evt_inline_c_0).booleanValue );
}


NSData * inline_c_Language_OSA_AppleEventDescriptor_1_331288a4ca5a9f60de3147162bff9373327b250a(NSAppleEventDescriptor * evt_inline_c_0) {
return (evt_inline_c_0.data);
}


int inline_c_Language_OSA_AppleEventDescriptor_2_9da62ad1cf17362296a544e60d1626fb6a37878c(NSAppleEventDescriptor * evt_inline_c_0) {
return ( (evt_inline_c_0).int32Value );
}


NSInteger inline_c_Language_OSA_AppleEventDescriptor_3_1051e0a914033f18b8a2597fd051ce2755312f69(NSAppleEventDescriptor * evt_inline_c_0) {
return ( (evt_inline_c_0).numberOfItems );
}


NSString * inline_c_Language_OSA_AppleEventDescriptor_4_e3101dcc161d8896ac668c10902eceab37a72cec(NSAppleEventDescriptor * evt_inline_c_0) {
return ( (evt_inline_c_0).stringValue );
}


NSAppleEventDescriptor * inline_c_Language_OSA_AppleEventDescriptor_5_3ee380eef71945efb604624a1f56ae6f0cb37464(NSAppleEventDescriptor * evt_inline_c_0, NSInteger n_inline_c_1) {
return ( [(evt_inline_c_0)
                                                      descriptorAtIndex: n_inline_c_1] );
}

File: src/Language/OSA/AppleEventDescriptor.hs
{-# LANGUAGE DataKinds, GADTs, OverloadedStrings, PolyKinds, QuasiQuotes  #-}
{-# LANGUAGE RecordWildCards, TemplateHaskell, TypeFamilies, ViewPatterns #-}
-- | Provides API to extract information from AppleEventDescriptor, not creating.
module Language.OSA.AppleEventDescriptor
       (NSAppleEventDescriptor,
        asBool, data_, asInt32, itemCount, asText,
        descriptorAt) where

import           Data.ByteString      (ByteString)
import           Data.Int             (Int32)
import           Data.Text            (Text)
import           Foreign              (nullPtr)
import           Foreign.C.Types      (CInt (..), CLong (..))
import           Language.ObjC.Inline (objcCtxWithClasses)
import           Language.ObjC.Inline (defClass)
import           Language.ObjC.Inline (import_)
import           Language.ObjC.Inline (ObjC (..))
import           Language.ObjC.Inline (fromObjC)
import qualified Language.ObjC.Inline as C

C.context (objcCtxWithClasses [defClass "NSAppleEventDescriptor"
                              ,defClass "NSAppleScript"
                              ,defClass "NSDictionary"])

import_ "Foundation/Foundation.h"

type NSAppleEventDescriptor = ObjC "NSAppleEventDescriptor"

asBool :: NSAppleEventDescriptor -> IO Bool
asBool evt = (1 ==) <$> [C.exp| BOOL { ($raw:(NSAppleEventDescriptor *evt)).booleanValue } |]

data_ :: NSAppleEventDescriptor -> IO ByteString
data_ evt = [C.exp'| NSData * {$raw:(NSAppleEventDescriptor *evt).data} |]

asInt32 :: NSAppleEventDescriptor -> IO CInt
asInt32 evt = [C.exp| signed int { ($raw:(NSAppleEventDescriptor *evt)).int32Value } |]

itemCount :: NSAppleEventDescriptor -> IO Int
itemCount evt = fromIntegral <$> [C.exp| NSInteger { ($raw:(NSAppleEventDescriptor *evt)).numberOfItems } |]

asText :: NSAppleEventDescriptor -> IO (Maybe Text)
asText evt = do
  txt@(ObjC obj) <- [C.exp| NSString * { ($raw:(NSAppleEventDescriptor *evt)).stringValue } |]
  if obj == nullPtr
    then return Nothing
    else Just <$> fromObjC txt

descriptorAt :: NSAppleEventDescriptor -> Int -> IO (Maybe NSAppleEventDescriptor)
descriptorAt evt (fromIntegral -> n) = do
  o@(ObjC ptr) <- [C.exp| NSAppleEventDescriptor * { [($raw:(NSAppleEventDescriptor *evt))
                                                      descriptorAtIndex: $(NSInteger n)] } |]
  if ptr == nullPtr
  then return Nothing
  else return $ Just o
